add winograd c4 implement (#2494)

MyPandaShaoxiang opened the pull request:
test=develop
fix: fix conv_block prepack_input_nxwc4 bug
* fix: optimize sgemm_c4 in armv7
     change condition of choose winograd kernel
* fix: change conv choose kernel condition
test=develop

MyPandaShaoxiang committed:
add winograd c4 implement (#2494)

fix: fix conv_block prepack_input_nxwc4 bug
* fix: optimize sgemm_c4 in armv7
     change condition of choose winograd kernel
* fix: change conv choose kernel condition

xyoungli commented on lite/kernels/arm/conv_compute.cc:
```diff
@@ -67,7 +68,15 @@ void ConvCompute<PRECISION(kFloat), PRECISION(kFloat)>::PrepareForRun() {
              no_dilation) {
-    if (ic >= 32 && oc >= 32 && hout > 16 && wout > 16) {
+    int tile_block = 8;
+#ifdef __aarch64__
```
tile block是否要用于计算并行度？

xyoungli commented on lite/kernels/arm/conv_winograd.cc:
```diff
@@ -36,77 +37,89 @@ void WinogradConv<PRECISION(kFloat), PRECISION(kFloat)>::ReInitWhenNeeded() {
-
-  const int n_wino = size_tile;
-  workspace_size_ = (size_trans_channel * max_ch * 2 + n_wino) * sizeof(float);
+  int oh = o_dims[2];
```
不需要每次都trans weights，如果选的kernel跟上次一致的话，不需要转

MyPandaShaoxiang committed:
fix winograd reinitwhenneed (#2511)


* add winograd c4 implement (#2494)
test=develop
fix: fix conv_block prepack_input_nxwc4 bug
* fix: optimize sgemm_c4 in armv7
     change condition of choose winograd kernel
* fix: change conv choose kernel condition
test=develop

xyoungli commented on lite/kernels/arm/conv_compute.cc:
```diff
@@ -67,7 +68,11 @@ void ConvCompute<PRECISION(kFloat), PRECISION(kFloat)>::PrepareForRun() {
-    if (ic >= 32 && oc >= 32 && hout > 16 && wout > 16) {
+    bool use_winograd =
+        (threads == 1 && oc >= 4 && ic >= 4 && hout >= 6 && wout >= 6 &&
+         pads_equal) ||
```
要求pads_equal吗？

MyPandaShaoxiang replied:
目前是需要的，代码没有对4pad做处理，后期会加上

xyoungli reviewed (approved):
LGTM
